### ISSUE TITLE
Golf: crash when at loading animation when fullscreen mode is active (Apple silicon)

### ISSUE THREAD

**Issue by gingerbeardman** [fallahn/crogine#67]
## golf

1. start game in fullscreen mode

or

1. start game in windowed mode
1. go fullscreen
1. do not forward any ports
1. attempt to create game
1. error on restart

Expected
- game should not crash

Actual
- game crashes

Affected
- custom macOS Apple silicon build

Not Affected
- itch macOS Intel build

**Comment by fallahn** [comment]
Hum, the fact that it doesn't happen on the intel version but (presumably) on the same hardware with a native version makes it tricky to debug without direct access. Is there a copy of the call stack in the error report when it crashes which you can share?

**Comment by gingerbeardman** [comment]
OK, I need to rephrase things... it hangs rather than crashes.

I can exit the game by pressing Cmd+Option+Esc.

## Screen Recording

https://user-images.githubusercontent.com/49612/221267333-09fec764-05b6-47a1-a93c-ab05bd298486.mov

**Comment by fallahn** [comment]
I see. It might still be a crash as the loading screen is running in its own thread while the main thread does the heavy lifting. If the main thread crashes the loading screen thread is left in a bad state, rather than exiting properly. You'll have to launch it via a debugger (and even better build it in debug mode) to see if you can catch any errors.

**Comment by gingerbeardman** [comment]
If it helps, this is the last line of the debug log.

```
2023-02-24 21:04:24.909 golf[91360:5647206] ERROR: Error(2) opening /Users/matt/Library/Application Support/Trederia/golf/courses/export/
```

**Comment by fallahn** [comment]
That should be a warning really (that's on me), rather than an error, as it only checks to see if there's a folder containing courses from the course editor. You can always create the directory yourself to see if the problem goes away, but I think it's just coincidental.

I don't know of any specific guides, as macs aren't really my area, but you ought to be able to open the XCode project generated with cmake (in the build dir) with the XCode IDE and debug it with that. That should be well documented at least. That way when you hit a crash/error it'll pop up with something a bit more meaningful in XCode rather than just pooping the bed.

**Comment by gingerbeardman** [comment]
Sorry what I mean that in the logical flow of the program, the crash happens after that line is printed.

Thinking of the old print/debug dance.

**Comment by gingerbeardman** [comment]
Sadly Xcode project gives me all manner of seemingly-nonsensical permissions errors when I try to run it or build it from there.

I will look into it.

**Comment by gingerbeardman** [comment]
OK, I am using Instruments and the thread is locking on a call to `sys_ulock_wait`

Backtrace details in the screenshot.

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/49612/221302840-16fbafd6-64ff-4a0a-9310-c17f976ea28c.png">

**Comment by fallahn** [comment]
Hmm.. a race condition maybe? Hard to tell. You could try disabling the loading screen altogether (though the game will appear to freeze while it loads). At the top of MenuState.cpp (around line 147 or so)

    context.mainWindow.loadResources([this]() {
        //add systems to scene
        addSystems();
        //load assets (textures, shaders, models etc)
        loadAssets();
        //create some entities
        createScene();
    });

can be changed to

    //context.mainWindow.loadResources([this]() {
        //add systems to scene
        addSystems();
        //load assets (textures, shaders, models etc)
        loadAssets();
        //create some entities
        createScene();
    //});

(comment out the top and bottom lines). You could also try adding some cout debugging between each function call just to see where it gets stuck.

**Comment by gingerbeardman** [comment]
I'm still getting this crash occaisionally even when not running full screen.

- custom Apple silicon build

**Comment by gingerbeardman** [comment]
Crashing:

- I can avoid the crash if i do not resize or move the window! Bizarre.

**Comment by fallahn** [comment]
[Possibly related](https://github.com/libsdl-org/SDL/issues/5203)?

**Comment by gingerbeardman** [comment]
Possibly related, but I am using latest SDL2 framework (2.26.4, from 2 weeks ago) and still get the crash.

**Comment by fallahn** [comment]
Fixed in [7f1a195](https://github.com/fallahn/crogine/pull/85/commits/7f1a1958d3dba49ca8f18662557f021099e6c89e)
